src/transformStateWithClones.js

```diff
@@ -7,7 +7,39 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    let newState;
+
+    switch (action.type) {
+      case 'clear':
+        newState = {};
+        break;
+
+      case 'addProperties':
+        newState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        newState = { ...currentState };
+
+        action.keysToRemove.forEach((key) => {
+          delete newState[key];
+        });
+        break;
+      default:
+        throw new Error(`Unknown action type: ${action.type}`);
```

Comment:
The `switch` statement is correctly implemented with a `default` case for error handling. However, make sure to handle the case where `action` might not have a `type` property, which could lead to an undefined behavior.